### PR TITLE
feature/15142-new-indicator-chande

### PR DESCRIPTION
**Pull request by pawellysy**
Enhancement #15142 Added new Indicator - Chande Momentum Oscilator..

### CODE REVIEW

File: ts/Stock/Indicators/CMO/CMOIndicator.ts [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    isNumber,
+    merge
+} = U;
+
+/* eslint-disable require-jsdoc */
+
+// Utils:
+function toFixed(a: number, n: number): number {
+    return parseFloat(a.toFixed(n));
+}
```
Comment (pawelfus) [inline ts/Stock/Indicators/CMO/CMOIndicator.ts:36]: This function seems to be unused

File: ts/Stock/Indicators/CMO/CMOOptions.d.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ * */
+
+export interface CMOOptions extends SMAOptions {
+    params?: CMOParamsOptions;
+}
+
+export interface CMOParamsOptions extends SMAParamsOptions {
+    // for inheritance
+    decimals?: number;
```
Comment (pawelfus) [inline ts/Stock/Indicators/CMO/CMOOptions.d.ts:32]: This option is not used anywhere

**Review by pawelfus — CHANGES_REQUESTED** [review]
Overall good! Minor cleanup changes

File: ts/Stock/Indicators/CMO/CMOIndicator.ts [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+    /* *
+     *
+     *  Properties
+     *
+     * */
+
+    public data: Array<CMOPoint> = void 0 as any;
+    public points: Array<CMOPoint> = void 0 as any;
+    public options: CMOOptions = void 0 as any;
```
Comment (karolkolodziej) [inline ts/Stock/Indicators/CMO/CMOIndicator.ts:79]: I would organize them alphabetically.

File: ts/Stock/Indicators/CMO/CMOIndicator.ts [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        }
+
+        return { sumOfLowerValues, sumOfHigherValues };
+    }
+
+    public getValues<TLinkedSeries extends LineSeries>(
+        series: TLinkedSeries,
+        params: CMOParamsOptions
+    ): IndicatorValuesObject<TLinkedSeries> | undefined {
+        var period = params.period as any,
```
Comment (karolkolodziej) [inline ts/Stock/Indicators/CMO/CMOIndicator.ts:111]: We might split them with `const/let` to better see which can be modified and which not.

**Review by karolkolodziej — CHANGES_REQUESTED** [review]
Minor changes/suggestions ;)

**Review by karolkolodziej — APPROVED** [review]

File: ts/Stock/Indicators/CMO/CMOIndicator.ts [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        }
+
+        if (isNumber(yVal[0])) {
+            values = yVal as Array<number>;
+        } else {
+            // in case of the situation, where the series type has data length
+            // shorter then 4 (HLC, range), this ensures that we are not trying
+            // to reach the index out of bounds
+            index = Math.min(index, yVal[0].length - 1);
+            values = (yVal as Array<Array<number>>).map((value: Array<number>): number => value[index]);
```
Comment (raf18seb) [inline ts/Stock/Indicators/CMO/CMOIndicator.ts:127]: Please, try not to extend the 80 characters line length

File: ts/Stock/Indicators/CMO/CMOIndicator.ts [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        }
+
+        for (i = period; i < yValLen; i++) {
+            const { sumOfLowerValues, sumOfHigherValues } = this.calculateSums(period, i, values),
```
Comment (raf18seb) [inline ts/Stock/Indicators/CMO/CMOIndicator.ts:131]: 80

File: ts/Stock/Indicators/CMO/CMOIndicator.ts [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+                x = xVal[i],
+                y = 100 * (sumOfHigherValues - sumOfLowerValues) / (sumOfHigherValues + sumOfLowerValues);
```
Comment (raf18seb) [inline ts/Stock/Indicators/CMO/CMOIndicator.ts:133]: 80

File: ts/Stock/Indicators/CMO/CMOIndicator.ts [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+            // in case of the situation, where the series type has data length
+            // shorter then 4 (HLC, range), this ensures that we are not trying
+            // to reach the index out of bounds
+            index = Math.min(index, yVal[0].length - 1);
+            values = (yVal as Array<Array<number>>).map((value: Array<number>): number => value[index]);
+        }
+
+        for (i = period; i < yValLen; i++) {
+            const { sumOfLowerValues, sumOfHigherValues } = this.calculateSums(period, i, values),
+                x = xVal[i],
```
Comment (raf18seb) [inline ts/Stock/Indicators/CMO/CMOIndicator.ts:132]: It's not a mistake, but in my opinion, I think that the `x` variable is not needed.

File: ts/Stock/Indicators/CMO/CMOIndicator.ts (hunk 6 above)
Comment (raf18seb) [inline ts/Stock/Indicators/CMO/CMOIndicator.ts:131]: Here you have 2 `for` loops. The first one is `for (i = period; i < yValLen; i++) {` that iterates `yValLen` times and, inside the `calculateSums` method, you have another `for` loop `for (let j = i; j > i - period; j--) {` that iterates `period` times. I think you could improve the performance drastically using only one loop `for (i = period; i < yValLen; i++) {` and calculating the sums "on fly".

File: ts/Stock/Indicators/CMO/CMOIndicator.ts (hunk 5 above)
Comment (raf18seb) [inline ts/Stock/Indicators/CMO/CMOIndicator.ts:127]: Here is another loop that iterates through all `yVal` values - it should be possible to put it all together inside only one loop.

File: ts/Stock/Indicators/CMO/CMOIndicator.ts [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+
+    public calculateSums(
+        period: number,
+        i: number,
+        values: Array<number>
+    ): { sumOfLowerValues: number; sumOfHigherValues: number } {
+        let sumOfLowerValues = 0,
+            sumOfHigherValues = 0;
+
+
```
Comment (raf18seb) [inline ts/Stock/Indicators/CMO/CMOIndicator.ts:89]: 2 lines break

**Review by raf18seb — CHANGES_REQUESTED** [review]
Except for a few minor things, I have one suggestion about how the performance can be improved (see comments).

**Comment by TorsteinHonsi** [comment]
Thanks, nice work! Awaiting:
- Merge in master to resolve conflicts
- Final approval of @raf18seb and @pawelfus

File: ts/Stock/Indicators/CMO/CMOIndicator.ts [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+ * @extends   series,plotOptions.cmo
+ * @since     next
+ * @product   highstock
+ * @excluding dataParser, dataURL
+ * @requires  stock/indicators/indicators
+ * @requires  stock/indicators/cmo
+ * @apioption series.cmo
+ */
+
+(''); // to include the above in the js output
```
Comment (pawelfus) [inline ts/Stock/Indicators/CMO/CMOIndicator.ts:214]: Question to @bre1470 - does it matter if we end with `('')` or `''`?

**Review by pawelfus — APPROVED** [review]
Approved 👍

File: ts/Stock/Indicators/CMO/CMOIndicator.ts (hunk 10 above)
Comment (bre1470) [reply ts/Stock/Indicators/CMO/CMOIndicator.ts:214]: It's cute! Looks like it got compressed out, so it should be okay. 👍

Comment (pawelfus) [reply ts/Stock/Indicators/CMO/CMOIndicator.ts:214]: I agree! :D Looks like a happy robot sending you thumbs up!
`('');` ->   ('')👍

**Review by raf18seb — APPROVED** [review]
I'm only concerned about missing `correctFloat` at +/- operations. But it's up to you @pawellysy and @TorsteinHonsi 
The rest seems to be OK.

**Review by TorsteinHonsi — APPROVED** [review]
> I'm only concerned about missing correctFloat at +/- operations

It also costs some performance, so I'd say we deal with it if problems arise.

**Pull request merged**